nzshm_common/location/coded_location.py

```diff
@@ -0,0 +1,329 @@
+        return self.downsample(resolution)
+
+    def downsample(self, resolution: float) -> "CodedLocation":
+        """
+        Create a downsampled CodedLocation with a coarser resolution.
+
+        Examples:
+            ```py
+            >>> from nzshm_common import location
+            >>> loc_akl = location.get_locations(["AKL"])[0]
+            >>> loc_akl
+            CodedLocation(lat=-36.87, lon=174.77, resolution=0.001)
+            >>> loc_akl.downsample(0.1)
+            CodedLocation(lat=-36.9, lon=174.8, resolution=0.1)
+            >>> loc_akl.downsample(0.5)
+            CodedLocation(lat=-37.0, lon=175.0, resolution=0.5)
+            ```
+        """
+        return CodedLocation(lat=self.lat, lon=self.lon, resolution=resolution)
+
+
+class CodedLocationBin:
+    """
+    A collection of CodedLocation values, gathered into bins at a coarser resolution.
+    """
+
+    reference_point: CodedLocation
+    locations: List[CodedLocation]
+    bin_resolution: float
+
+    def __init__(
+        self, reference_point: CodedLocation, bin_resolution: float, locations: Optional[Iterable[CodedLocation]] = None
+    ):
+        """
+        Create a CodedLocationBin instance.
+
+        Arguments:
+            reference_point: the downsampled coordinate use as a reference point for the collection.
+            bin_resolution: the coarser-level resolution of the bin
+            locations: a collection of CodedLocation values
+        """
+        self.reference_point = reference_point
+        self.bin_resolution = bin_resolution
+
+        if locations is not None:
+            self.locations: List[CodedLocation] = list(locations)
+        else:
+            self.locations: List[CodedLocation] = list()
+
+    def __iter__(self):
+        """Iterate over the location collection."""
+        return self.locations.__iter__()
+
+    def __len__(self):
+        """Count the number of locations in the bin."""
+        return len(self.locations)
+
+    def __repr__(self):
+        plural = "" if len(self.locations) == 1 else "s"
+        count = len(self)
+        code = self.reference_point.code
+        resolution = self.bin_resolution
+
+        return f"CodedLocationBin({count} location{plural} near {code} below resolution {resolution})"
+
+    @property
+    def code(self) -> str:
+        """
+        The string code for the reference point, expressed as "latitude~longitude".
+
+        This can be used as a unique key when accessing a dictionary of bins.
+
+        Example:
+            ```py
+            >>> my_bin = list(bins.values())[0]
+            >>> my_bin
+            CodedLocationBin(12 locations near -37.0~175.0 below resolution 0.5)
+            >>> bins[my_bin.code] == my_bin
+            True
+            ```
+        """
+        return self.reference_point.code
+
+
+def bin_locations(
+    locations: Iterable[CodedLocation], at_resolution: float, sort_bins: bool = True
+) -> OrderedDict[str, CodedLocationBin]:
```

Comment:
can this be Dict? I thought dicts since 3.7 were ordered

Reply:
Dictionaries since 3.7 [preserve the order of entry](https://mail.python.org/pipermail/python-dev/2017-December/151283.html) into the dict. So we get this behaviour:
```py
>>> s = {1:1, 2:2, 4:4}
>>> s
{1: 1, 2: 2, 4: 4}
>>> s[3]=3
>>> s
{1: 1, 2: 2, 4: 4, 3: 3}
```

I'm using an `OrderedDict` here because having an ordered collection of entries is the explicit intent; in a `dict` the ordering behaviour would only be _implicitly_ there.

An `OrderedDict` is considered a subtype of `Dict`, and should be able to behave as a Dict would in most cases. (The main difference is the string representation.)

But if you like, it can be converted to a native dictionary with `dict(bin_dict)` and made available as the broader type.

Reply:
Better to be explicit about the intent, keep as `OrderedDict`